rlog/2024-04-20-rln-light-verifiers.mdx

```diff
@@ -0,0 +1,173 @@
+---
+title: 'Verifying RLN proofs in light clients with Subtrees'
+date: 2024-04-20 12:00:00
+authors: p1ge0nh8er
+published: true
+slug: rln-light-verifiers
+categories: research
+
+toc_min_heading_level: 2
+toc_max_heading_level: 4
+---
+
+How resource-restricted devices can verify RLN proofs fast and efficiently.
+
+<!--truncate-->
+
+## Introduction
+
+Recommended previous reading: [Strengthening Anonymous DoS Prevention with Rate Limiting Nullifiers in Waku](https://vac.dev/rlog/rln-anonymous-dos-prevention).
+
+This post expands upon ideas described in the previous post,
+focusing on how resource-restricted devices can verify RLN proofs fast and efficiently.
+
+Previously, it was required to fetch all the memberships from the smart contract,
+construct the merkle tree locally,
+and derive the merkle root,
+which is subsequently used to verify RLN proofs.
+
+This process is not feasible for resource-restricted devices,
+since it involves a lot of RPC calls, computation and fault tolerance.
+One cannot expect a mobile phone to fetch all the memberships from the smart contract and construct the merkle tree locally.
+
+## Constraints and requirements
+
+An alternative solution to the one proposed in this post is to construct the merkle tree on-chain,
+and have the root accessible with a single RPC call.
+However, this approach increases gas costs for inserting new memberships,
+and _may_ not be feasible until it is optimized further,
+with batching mechanisms, etc.
+
+The other methods have been explored in more depth [here](https://hackmd.io/@rymnc/rln-tree-storages).
+
+Following are the requirements and constraints for the solution proposed in this post:
+
+1. Cheap membership insertions.
+2. As few RPC calls as possible to reduce startup time.
+3. Merkle root of the tree is available on-chain.
+4. No centralized services to sequence membership insertions.
+5. Map inserted commitments to the block in which they were inserted.
+
+## Metrics on Sync Time for a tree with 2653 leaves
+
+The following metrics are based on the current implementation of RLN in the Waku gen0 network.
+
+### Test bench
+
+- Hardware: Macbook Air M2, 16GB RAM
+- Network: 120 Megabits/sec
+- Nwaku Commit: [e61e4ff](https://github.com/waku-org/nwaku/tree/e61e4ff90a235657a7dc4248f5be41b6e031e98c)
+- RLN Membership set contract: [0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4](https://sepolia.etherscan.io/address/0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4#code)
+- Deployed block number: 4,230,716
+- RLN Membership set depth: 20
+- Hash function: PoseidonT3 (which is a gas guzzler)
+- Max size of the membership set: 2^20 = 1,048,576 leaves
+
+### Metrics
+
+- Time to sync the whole tree: 4 minutes
+- RPC calls: 702
+- Number of leaves: 2653
+
+One can argue that the time to sync the tree at the current state is not _that_ bad.
+However, the number of RPC calls is a concern,
```

Comment:
Consider changing to:
'However, the number of RPC calls, which scales linearly with the number of blocks since the contract was deployed, is a concern.'

Reply:
addressed in 6677c3d